Add documentation about auth and certificates

DirectXMan12 opened the pull request:
This adds a description of how the authentication and
authorization works for the service catalog API server, as well as how
to generate the appropriate certificates required for serving the
service catalog API, and for authentication.

DirectXMan12 commented:
cc @pmorie @arschles PTAL and let me know if this is readable to ~~real human beings~~ people who haven't been immersed in generic API server work for a while

MHBauer commented on docs/auth.md:
```diff
@@ -0,0 +1,191 @@
+
+## Quickstart
+
+1. Ensure that you have a ConfigMap in the `kube-system` namespace named
```
Maybe add the kubectl cmd? 
`kubectl get configmap --namespace kube-system extension-apiserver-authentication -o yaml` ?

MHBauer commented on docs/auth.md:
```diff
@@ -0,0 +1,191 @@
+   `apiserver.tls.requestHeaderCA` Helm chart option..
+4. If not running on the cluster, ensure that you pass in a Kubeconfig
+   file to authenticate with the main cluster using the
+   `--authentication-kubeconfig` and `--authorization-kubeconfig`.
```
These flags are passed to what?

MHBauer replied:
These are files or file contents?

MHBauer commented on docs/auth.md:
```diff
@@ -0,0 +1,191 @@
+(you can install `cfssl` using `go get -u
+github.com/cloudflare/cfssl/cmd/...`).
+
+In the common case, you'll want to re-use the CA certificates used by the
```
How do I do this reuse?

MHBauer commented on docs/auth.md:
````diff
@@ -0,0 +1,191 @@
+```
+
+This will generate a certificate and private key for the CA, as well as
+a signing configuration used by `cfssl` below. `$PURPOSE` may be set to
````
Move the ${PURPOSE} discussion above the commands because we need it for them to work.

Alternatively, leave it, so when people copy and paste it breaks. But they could stop and file an issue rather than read on.

MHBauer commented on docs/auth.md:
````diff
@@ -0,0 +1,191 @@
+using the following command:
+
+```shell
+$ openssl -x509 -sha256 -new -nodes -days 365 -newkey rsa:2048 -keyout ${PURPOSE}-ca.key -out ${PURPOSE}-ca.crt -subj "/CN=ca"
````
friendly reminder of 
`openssl req -x509`

DirectXMan12 commented:
updated for clarity post conversation with @MHBauer (thanks a ton! 😄 ), edited for compliance with the Kubernetes documentation style guide.

MHBauer commented on docs/auth.md:
```diff
@@ -0,0 +1,259 @@
+Additionally, it provides instructions for generating the serving
+certificates used to securely serve the service catalog API.
+
+## Certificates Overview
```
This section is really great and illuminating.

If there is relevance to k8s at large, I would hate for it to be stuck in our repo.

MHBauer commented on docs/walkthrough.md:
````diff
@@ -53,7 +53,14 @@ the appropriate values will be automatically populated.
+`apiserver.tls.requestHeaderCA` option when installing the Helm chart:
+
+```shell
+helm install charts/catalog --name catalog --namespace catalog --set apiserver.tls.requestHeaderCA=$(base64 --wrap 0 /path/to/requestheader-client-ca.crt)
````
oh nice, I totally forgot about using the shell like this.

arschles replied:
2 things:

1. We need a series of CLI commands in the walk-through to explain how to create one of these CAs, separate from the `auth.md` file
2. The auth.md file is great for more details, but from where do we link it?

MHBauer commented on docs/auth.md:
```diff
@@ -0,0 +1,259 @@
+
+This generates a certificate and private key for the CA, as well as
+a signing configuration used by `cfssl` below. `$PURPOSE` should be set to
+one of `serving`/`server`, `client`, or `requestheader-client`, as
```
I can verify the CA part plus the base64 part in two paragraphs works for getting past #737

The serving certificate stuff requires some testing. We should probably try and get some e2e on it.

MHBauer commented on docs/auth.md:
```diff
@@ -0,0 +1,259 @@
+Kubernetes API server.
+
+Generally, the default admin user in a cluster connects with client
+certificate authentication.  Additionally, off-cluster non-human clients
```
I like the phrase "non-human clients".

derekwaynecarr commented on docs/auth.md:
```diff
@@ -0,0 +1,259 @@
+
+The Kubernetes documentation has a [detailed
+section](https://kubernetes.io/docs/admin/authentication/#creating-certificates)
+on how to create certificates several different ways.  For convinience,
```
Not typo on convenience

derekwaynecarr commented on docs/auth.md:
```diff
@@ -0,0 +1,259 @@
+In order to securely serve the service catalog API server over HTTPS,
+you'll need serving certificates.  By default, a set of self-signed
+certificates are generated by the service catalog API server. However,
+client have no way to trust these (since they are self-signed, there is no
```
clients

DirectXMan12 commented on docs/auth.md:
```diff
@@ -0,0 +1,259 @@
+
+### Delegated Token Authentication
+
+Delegated token authentication authenticates clients who pass in a token
```
a couple things describe similar things, but, IIRC none of the Kubernetes docs actually describe *delegating* token auth.  I'll see what I can dig up.

DirectXMan12 commented:
This should fix the nits.  I've added a couple sentences on *self-signed* certificates, and added a link from the delegating token authentication section to the webhook token authentication documentation, which works very similarly in principal (but uses different flags).

arschles commented:
@DirectXMan12 is this supposed to fix #737?

krancour commented:
As I've said in #737, this is all enlightening, _but_ this is also really a ton of understanding and pre-reqs to expect someone to get through before they can install the chart just to kick the tires. We should really ensure that the insecure endpoint always works, if enabled, per the chart (which is currently the default).

DirectXMan12 commented:
Ok, so, as I noted in #737, I missed an option the first time around.  There is an option that is an "easy fix" for this, but it's not really a fix.  Passing it will effectively disable both client certificate authentication and requestheader authentication.  I've added an option to use this to the Helm charts, and updated the documentation accordingly.

DirectXMan12 commented:
After some further conversations with users and @pmorie, I've updated the Helm chart to just disable authentication and authorization by default (note that running the API server binary not through the helm chart still has auth enabled by default -- the helm chart simply passes the flag to disable by default).

This should probably be revisited once most deployment methods have the correct options set on the API server by default.

MHBauer commented:
Do we eventually want to have all of this on by default? What is the timeframe for that? Do we need a tracking issue?

DirectXMan12 commented:
> Do we eventually want to have all of this on by default? What is the timeframe for that?

Probably?

Basically, it comes down to: what portion of users do we feel comfortable telling to go read the docs because their deployment mechanism for their cluster didn't automatically set things up (i.e. didn't set the `--requestheader-client-ca-file` flag on the main Kubernetes API server)?

If we don't feel comfortable telling that to anyone, then we probably never turn it on by default.  If we want to wait until minikube supports it, then that's the gating factor, etc.

> Do we need a tracking issue?

Probably not a bad idea if we do decide we want to turn this on by default eventually.

pmorie commented:
> Do we eventually want to have all of this on by default? What is the timeframe for that? Do we need a tracking issue?

Yes, I think that we should turn it on default when delegated auth is expected to work in most clusters.  Right now, we know the minikube doesn't generate some of the artifacts that are required.  A ton of people use minikube, so I'd like to (for example) make sure that minikube works correctly before making auth the default.

vaikas-google commented on docs/auth.md:
```diff
@@ -0,0 +1,287 @@
+proxies, which themselves have already authenticated the client.  It works
+similarly to [client certificate
+authentication](#client-certificate-authentication): it validates the
+verifies the certificate of the proxy using a CA certificate.  However, it
```
...it validates the verifies the certificate... sounds odd.
...it validates the certificate?

vaikas-google commented on docs/auth.md:
```diff
@@ -0,0 +1,287 @@
+used for the main Kubernetes API server, and in the same format (e.g.
+Kubernetes RBAC).
+
+By default, the service catalog API server searches for the connect
```
...for the connection information?

vaikas-google commented on docs/walkthrough.md:
```diff
@@ -44,16 +44,16 @@ To install the service catalog system with sensible defaults:
-`apiserver.tls.requestHeaderCA` option when installing the Helm chart.
+**Note**: Authentication and authorization are disabled in the Helm chart
+by default, even on the "secure" endpoint.  To enable them, set the
+`apiserver.enableAuth` option on the Helm chart:
```
You can also do this by updating the values file?

krancour commented on charts/catalog/README.md:
```diff
@@ -54,6 +54,7 @@ chart and their default values.
 | `apiserver.storage.etcd.servers` | If storage type is `etcd`: etcd URL(s); override this if NOT using embedded etcd | `http://localhost:2379` |
 | `apiserver.storage.tpr.globalNamespace` | If storage type is `tpr`: Some service catalog resources are not namespaced, but third party resources must be; setting this designates a namespace that will be treated as a container for such resources | `servicecatalog` |
 | `apiserver.verbosity` | Log level; valid values are in the range 0 - 10 | `10` |
+| `apiserver.enableAuth` | Enable authentication and authorization for the service catalog API server | false |
```
I would put backticks around false since we're speaking literally about a boolean value.

krancour replied:
Can also drop "for the service catalog API server". This bit of info is redundant given that the attribute in question is already qualified by `apiserver.`

krancour replied:
You can... but in case it's not obvious, hacking this chart's (or any chart's) included `values.yaml` isn't the _right_ way to override values. (But I'm not saying I never do it. 😉) The _right_ way (apart from using `--set`) is to provide your _own_ YAML file with overrides and point to it using `--values`. I don't think it's 100% necessary to document that here (i.e. no action required imo), but if we _do_ want to make any mention of file-driven configuration of a chart, it shouldn't encourage updating the existing `values.yaml`. One obvious danger or that is contributors who edit it can easily, accidentally submit local changes they did not intend to along with their next PR.

pmorie replied:
> but if we do want to make any mention of file-driven configuration of a chart, it shouldn't encourage updating the existing values.yaml.

agree

krancour commented on charts/catalog/values.yaml:
```diff
@@ -44,6 +44,9 @@ apiserver:
   verbosity: 10
+  # disable authentication so that people don't have to bother setting up certificates
+  # properly just to get the walkthrough working
+  enableAuth: false
```
Two nits here. To be consistent with the rest of the comments in the `values.yaml` file, we should upcase the leading letter of this comment. Also, for consistency, we should probably be documenting the function of the switch instead of the default _value_ of the switch. i.e. Instead of saying "disable authentication ..." (which documents the default value), say instead "Enable or disable authentication ... Disabling can be useful for kicking the tires in a local cluster, but is not suitable for production."

krancour replied:
Having written this Helm chart and others... a lesson I have learned is to plan for the future. If you think there will be _more_ authentication-related settings we will want to concern ourselves with in the future (which I think is very likely), we should consider adopting this structure instead:

```
apiserver:
  ...
  auth:
    # Comment...
    enabled: false
```

This allows room for growth without having to make breaking changes or resorting to uglier attributes that would look like `apiserver.authOptionX` and `apiserver.authOptionY`.

krancour commented on docs/auth.md:
```diff
@@ -0,0 +1,287 @@
+# Serving Certificates, Authentication, and Authorization
```
I'm very happy to see the walkthrough has been updated to mention the Helm chart disables authn/authz by default, but could we also mention the same toward the top of this `auth.md` document so that anyone who's just kicking the tires won't waste too much time reading over this document. (Please don't mistake that a me undervaluing this document. This is _tremendously_ important for people who intend to use this in production and you have done a fantastic job of explaining things.)

krancour commented:
I'm very happy with this, overall, but I did leave a few new comments-- all of which, I believe can be quite easily addressed. I'd like to see those done before LGTM3 and merge.

DirectXMan12 commented:
@vaikas-google @krancour All your comments should be addressed

krancour commented on charts/catalog/values.yaml:
```diff
@@ -44,6 +44,11 @@ apiserver:
+    # Enable or disable authentication and authorization.  Disabling authentication
+    # and authorization can be useful for quickly getting the walkthrough up and running,
+    # but is not suitable for production.
+    enable: false
```
Nit: I would prefer that this were `enabled`.

krancour commented:
@DirectXMan12 this looks awesome. I registered [one final nit](https://github.com/kubernetes-incubator/service-catalog/pull/745/files#r113771820). It's up to you whether you want to act on it or not. If you do, LGTM, if not, lmk and I will not withhold my LGTM.

DirectXMan12 commented:
alright, the nit should be taken care of.

krancour commented:
I see you updated accordingly already. LGTM3

krancour commented:
Will merge pending CI success.
